src/OrleansRuntime/MultiClusterNetwork/MultiClusterOracle.cs

```diff
@@ -0,0 +1,795 @@
+                {
+                    result.Add(silo);
+                }
+            }
+
+            logger.Verbose("--- FindLaggingSilos: done, found {0}", result.Count);
+
+            return result;
+        }
+
+        private void PublishMyStatusToNewDestinations(MultiClusterData delta)
+        {
+            // for quicker convergence, we publish active local status information
+            // immediately when we learn about a new destination
+
+            GatewayEntry myEntry;
+
+            if (!localData.Current.Gateways.TryGetValue(this.Silo, out myEntry)
+                || myEntry.Status != GatewayStatus.Active)
+                return;
+
+            foreach (var gateway in delta.Gateways.Values)
+            {
+                GossipStatus gossipStatus;
+                var destinationCluster = gateway.ClusterId;
+
+                if (destinationCluster == this.clusterId)
+                {
+                    // local cluster
+                    this.siloTasks.TryGetValue(gateway.SiloAddress, out gossipStatus);
+                    if (!this.siloTasks.TryGetValue(gateway.SiloAddress, out gossipStatus) || !gossipStatus.KnowsMe)
+                        PublishGossipToSilo(gateway.SiloAddress, new MultiClusterData(myEntry));
+                }
+                else
+                {
+                    // remote cluster
+                    if (!this.clusterTasks.TryGetValue(destinationCluster, out gossipStatus) || !gossipStatus.KnowsMe)
+                        PublishGossipToCluster(destinationCluster, new MultiClusterData(myEntry));
+                }
+            }
+        }
+
+        private class GossipStatus
+        {
+            public SiloAddress Address;
+            public Task Task = TaskDone.Done;
+            public Exception LastException;
+            public bool KnowsMe;
+            public bool Pending;
+            public DateTime LastUse = DateTime.UtcNow;
+        }
+
+        // tasks for gossip
+        private readonly Dictionary<SiloAddress, GossipStatus> siloTasks = new Dictionary<SiloAddress, GossipStatus>();
+        private readonly Dictionary<string, GossipStatus> clusterTasks = new Dictionary<string, GossipStatus>();
+        private readonly Dictionary<IGossipChannel, GossipStatus> channelTasks = new Dictionary<IGossipChannel, GossipStatus>();
+ 
+        // numbering for tasks (helps when analyzing logs)
+        private int idCounter;
+
+        private void PublishGossipToSilo(SiloAddress destinationSilo, MultiClusterData delta)
+        {
+            GossipStatus status;
+            if (!this.siloTasks.TryGetValue(destinationSilo, out status))
+                this.siloTasks[destinationSilo] = status = new GossipStatus();
+
+            int id = ++this.idCounter;
+            logger.Verbose("-{0} PublishGossipToSilo {1} {2}", id, destinationSilo, delta);
+
+            // the task that actually publishes
+            Func<Task, Task> publishAsync = async (Task prev) =>
```

Comment:
Yes, this is not perfect. I will improve it, e.g. using the same pattern I use for the QueuedGrain (when batching writebacks to storage).